opencensus/log/__init__.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2019, OpenCensus Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import namedtuple
+from copy import copy
+import logging
+
+from opencensus.trace import execution_context
+
+
+_meta_logger = logging.getLogger(__name__)
+
+TRACE_ID_KEY = 'traceId'
+SPAN_ID_KEY = 'spanId'
+SAMPLING_DECISION_KEY = 'traceSampled'
+
+LogAttrs = namedtuple('LogAttrs', ['trace_id', 'span_id', 'sampling_decision'])
+ATTR_DEFAULTS = LogAttrs('', '', False)
+
+
+def get_log_attrs():
+    """Get logging attributes from the opencensus context.
+
+    :rtype: :class:`LogAttrs`
+    :return: The current span's trace ID, span ID, and sampling decision.
+    """
+    try:
+        tracer = execution_context.get_opencensus_tracer()
+        if tracer is None:
+            raise RuntimeError
+    except Exception:  # noqa
+        _meta_logger.error("Failed to get opencensus tracer")
+        return ATTR_DEFAULTS
+
+    try:
+        trace_id = tracer.span_context.trace_id
+        if trace_id is None:
+            trace_id = ATTR_DEFAULTS.trace_id
+    except Exception:  # noqa
+        _meta_logger.error("Failed to get opencensus trace ID")
```

Comment:
I wish to understand more about the design here.

Do we expect it to be a very rare case when there is no trace_id/span_id?
For example, if a batch application is fetching items from a queue, and process the queue items. I would expect the item processing logic to have trace_id/span_id/etc., while the app initialization and queue fetching logic could be context free. Would this generate too many error log entries?

Reply:
I'd expect it to be rare that we don't have a span ID, and that we should always have a trace ID if OC is configured correctly. The broad exception handling here is because I really don't want an opencensus error to break application logging.

What do you think about adding config to attach a `NullHandler` to the `opencensus.log` logger so users don't see these errors unless they opt in?

Reply:
That sounds okay to me.
Do we consider putting these logs at `debug level`?

Reply:
I think `warn` is justified here if users have to specifically enable this logger, since I wouldn't expect to have to add a handler and set a lower log level to see these lines as a user.

Looking at https://docs.python.org/3.7/howto/logging.html#library-config, we've got three options:
1. don't set a default handler: `warn` and higher logs to stderr unless users configure a handler of their own
2. attach a `NullHandler` to the `opencensus.log` logger: users won't see logs at any level unless they configure their own handler
3. attach a `NullHandler` to the `opencensus` logger: same as (2) but applies to the whole package, users may enable logging for specific subpackages

We're doing (1) now, which seems like a sane default, but (3) might be the best long-term option.

Reply:
Makes sense.

Reply:
Particular integrations may not be enabled, but we should always get at least a trace ID here since we create a default tracer and span context on the first call to `get_opencensus_tracer`.